flocker/node/agents/gce.py

```diff
@@ -35,6 +35,68 @@
 _METADATA_HEADERS = {u'Metadata-Flavor': u'Google'}
 
 
+def wait_for_operation(compute, operation, timeout):
+    """
+    Blocks until a GCE operation is complete, or timeout passes.
+
+    This function will then poll the operation until it reaches state
+    'DONE' or times out, and then returns the final operation resource
+    dict.
+
+    :param operation: A dict representing a pending GCE operation resource.
+        This can be either a zone or a global operation.
+    :param timeout: The amount of times in seconds to wait until timing out the
+        operation.
+
+    :returns dict: A dict representing the concluded GCE operation
+        resource or `None` if the operation times out.
+    """
+    operation_name = operation['name']
+    if 'zone' in operation:
+        # For zone operations, we need to extract the zone and the project.
+        # An example of a 'zone' value of a zone operation dict is:
+        #
+        # 'https://content.googleapis.com/compute/v1/'
+        # 'projects/<project-name>/zones/<zone-name>'
+        #
+        # Note that everything before 'projects' is sometimes not included.
+        zone_url_parts = operation['zone'].split('/')
+        project = zone_url_parts[-3]
+        zone = zone_url_parts[-1]
```

Comment:
:question: unpacking these from a URL isn't very pleasant. From a quick look it
seems most callers know these values, could they be passed in, perhaps with
two functions: one for global, one for zone operations?

If 'status' == 'DONE' is standard across all GCE operations, then something
that loops, checks responses for that, and handles the timeouts, delays etc.
seems good.

If feels like it could be composed with functions to create these update checks,
and perhaps one to check the output for success, and avoid having these URL
operations.

I'd be ok with what we have here though, it seems GCE is consistent enough
that we won't have a problem with it.